charms/worker/k8s/src/charm.py

```diff
@@ -234,12 +237,49 @@ def _create_cos_tokens(self):
     @on_error(
         WaitingStatus("Waiting for enable components"), InvalidResponseError, K8sdConnectionError
     )
-    def _enable_components(self):
+    def _configure_components(self):
         """Enable necessary components for the Kubernetes cluster."""
-        status.add(ops.MaintenanceStatus("Enabling DNS"))
-        self.api_manager.enable_component("dns", True)
+        if self._dns_charm_integrated():
+            status.add(ops.MaintenanceStatus("Disabling DNS"))
+            self.api_manager.configure_component("dns", False)
+            self._configure_dns()
+        else:
+            status.add(ops.MaintenanceStatus("Enabling DNS"))
+            self.api_manager.configure_component("dns", True)
```

Comment:
Might take a more integrated approach here -- where `dns_settings` is some `Optional[DNSSettings]`

if dns_settings is `None`, `not None` is `True`, but if it's something, then it will be False

```suggestion
        dns_settings = self._dns_charm_integrated()
        status.add(ops.MaintenanceStatus(f"Configuring DNS"))
        self._configure_dns(external)
```